update setup procedures for LADAS

saraqzhang opened the pull request:
for integrated setup of adas and ldas, modifications and updates are implemented in fvsetup and several scripts and rc files  that are used in the fvsetup procedure.  the modifications to other related external components and ldas are updated by pull requests to the corresponding repositories. 

Related PR: https://github.com/GEOS-ESM/GEOSldas_GridComp/pull/81 (merged; follow-up PR in prep)

Requires:
https://github.com/GEOS-ESM/GEOSgcm_App/pull/689 (merged)
https://github.com/GEOS-ESM/GMAO_Shared/pull/370 (merged)

saraqzhang committed:
updates for integrated ladas setup procedure

saraqzhang committed:
add GID and atmens-replay-check to ldas_setup options

saraqzhang commented:
@rtodling  @gmao-jstassi @gmao-rreichle   This PR ( draft)  contains the updates we presented last meeting (1/22/2025). these updates have been tested in a hybrid 4dnvar experiment ( integrated setup and one-day cycling) .  Next we are going to implement the update to make lfo forcing as bkg files in the setup, workflow and archive. There will be corresponding development and updates in ldasGC src.

gmao-rreichle committed:
removed Haswell & Skylake choices, fixed typo (x3dvar_ladas.input)

gmao-rreichle commented on src/Applications/GEOSdas_App/testsuites/x3dvar_ladas.input:
```diff
@@ -44,7 +44,7 @@ FVHOME? [/discover/nobackup/qzhang/x3dvar_ladas]
 > 
 
-Processing nodes (1:Haswell, 2:Skylake, 3:Cascase, 4:Milan)? [2]
+Processing nodes (3:Cascade, 4:Milan)? [3]
```
this may need further tweaks.  maybe milan should be the default

saraqzhang replied:
this example  3dvar setup input file should be replaced by a hybrid4dnvar input file based on x050 setup.

gmao-rreichle replied:
FYI, I just merged the latest develop into the PR's branch, which required resolving a merge conflict for the above lines.  I'm a bit unsure about the defaults, but I'm fairly confident it doesn't matter at this moment.  I mostly wanted to make sure the branch still builds after merging the latest develop

gmao-rreichle commented on src/Applications/GEOSdas_App/GEOSdas.csm:
```diff
@@ -6082,9 +6082,9 @@ endif
+    /bin/rm -f $FVHOME/recycle/holdpredout/*lfo_Nx+-*
     /bin/mv -f $FVWORK/*lfo_Nx+-* $FVHOME/recycle/holdpredout/. 
-    /bin/cp $FVHOME/recycle/holdpredout/*lfo_Nx+-*  $LDHOME/input/met_forcing/. 
+    /bin/cp $FVHOME/recycle/holdpredout/*lfo_Nx+-*  $LDHOME/input/met_forcing/.
```
Why should we merge "Nx+-" now, when we're already committed to switching to a "bkg" HISTORY Collection?

saraqzhang replied:
lfo_Nx+-  is replaced by *bkg.*_lfo*,  handled by adas scripts in setup and during cycling,  no longer copied to LDHOME.

gmao-rreichle commented on src/Applications/NCEP_Etc/NCEP_enkf/scripts/gmao/etc/HISTAENS.rc.tmpl:
```diff
@@ -52,6 +52,8 @@ COLLECTIONS: 'bkg.eta'
 #            'inst3_3d_wxme_Np'
 #            'tavg3_2d_wxme_Nx'
 # Land output
+#            'tavg1_2d_lfo_Nx+-'
```
See comment above -- it's probably better to make the switch to the "bkg" HISTORY "lfo" collection now

saraqzhang replied:
tavg1_2d_lfo_Nx+-  is  replaced by  bkg.tavg1_2d_lfo ,  so is inst1

gmao-rreichle reviewed (commented):
@saraqzhang : I added a very minor commit (removing Haswell and Skylake choices) and a couple of inline comments

saraqzhang committed:
Merge branch 'develop' into feature/saraqzhang/updatesetup4ladas

saraqzhang committed:
add bkglforst tar file to gen_silo_arc.pl

saraqzhang committed:
update for LADAS in GEOSdas_App/

saraqzhang committed:
update for LADAS in NCEP_enkf/scripts/gmao/

saraqzhang committed:
update for LADAS in NCEP_enkf/scripts/gmao/etc

saraqzhang commented:
@gmao-rreichle @gmao-jstassi @rtodling  
we added the further updates for land forcing "lfo" as bkg in setup and in hybrid da cycling. The updated system passed the tests on setup and a hybrid LADAS cycling.

rtodling commented on src/Applications/NCEP_Etc/NCEP_enkf/scripts/gmao/get_atmens_rst.pl:
```diff
@@ -24,11 +24,12 @@
 #-------------
 {
-    my ($atmens_date3_dir, $atmens_date_dir, $atmens_ebkg_dir);
+    my ($atmens_date3_dir, $atmens_date_dir, $atmens_ebkg_dir, $atmens_ebkglfo_dir);
```
@saraqzhang I am puzzled about the changes here ... do these mean that now we have to have RSTs for LDAS regardless of whether ldas is running or not?

saraqzhang replied:
yes, we wanted to have bkglfo produced and archived regardless ldas.  so that we have the tar ball to be installed with other ebkg  by fvsetup later for coupled experiments.

gmao-rreichle replied:
> we have to have RSTs for LDAS regardless of whether ldas is running or not?

Ricardo, just to emphasize, the "bkglfo" files aren't strictly "restarts" from the perspective of the land system.  They're surface met forcing for the LDAS, and therefore more like time-varying boundary conditions.  It's useful to have this output for LDAS science development, even if the DAS isn't running the LDAS component.   If the burden of this extra output is too great, we can discuss.

gmao-rreichle committed:
Merge branch 'develop' into feature/saraqzhang/updatesetup4ladas

gmao-rreichle commented on src/Applications/GEOSdas_App/fvsetup:
```diff
@@ -3158,11 +3157,9 @@ sub get_nodeflg {
-  elsif ($ans == 2) { $nodeflg = "sky";  $ncpus_per_node = 36 }
-  elsif ($ans == 3) { $nodeflg = "cas";  $ncpus_per_node = 45 }
+  if    ($ans == 3) { $nodeflg = "cas";  $ncpus_per_node = 45 }
   elsif ($ans == 4) { $nodeflg = "mil";  $ncpus_per_node = 54 }
```
Is there a reason to use only 54 cores on mil?  Aren't there ~126 usable cores on each mil node?

saraqzhang replied:
leave this to @rtodling  and @gmao-jstassi  to make the call.

gmao-rreichle commented on src/Applications/GEOSdas_App/fvsetup:
```diff
@@ -5201,7 +5211,7 @@ EOF
        $ny = $nx * 6;
        $km = $vres;
-       $dt = 450;
+       $dt = 300;
```
This change of dt (heartbeat?) seems unrelated to LDAS.  Is this intentional?

saraqzhang replied:
this change was carried from adas5.41.0 ( likely ahead of develop for this modification), it is unrelated to LDAS. will consult with Ricardo and Joe about this.

gmao-rreichle commented on src/Applications/GEOSdas_App/fvsetup:
```diff
@@ -9781,12 +9795,13 @@ sub init_agcm_rc {
 
-    $flags{"gocart_tracers"} = $gocart_tracers;
     $flags{"carma_tracers"}  = $carma_tracers;
+    $flags{"gocart_tracers"} = $gocart_tracers;
```
Are we intentionally changing the order of the two lines here (relative to that in develop)?  Unless there's a reason for the revised order, we shouldn't modify develop

saraqzhang replied:
the order of the two lines are not changed in the code, not sure why it is indicated here.

gmao-rreichle replied:
?? I do see the change in the order of the lines when I do a direct (two-dot) comparison of develop vs. the PR's branch, see image below
<img width="2095" alt="Screenshot 2025-03-25 at 17 18 23" src="https://github.com/user-attachments/assets/2e78fc49-4394-4254-bb3d-a5e3db3ca45e" />

saraqzhang replied:
will make the order of the lines same as in develop ( different order was carried over in PR from 5.41. )

gmao-rreichle commented on src/Applications/GEOSdas_App/fvsetup:
```diff
@@ -9781,12 +9795,13 @@ sub init_agcm_rc {
     $flags{"iau"}            = $doiau;
-    $flags{"pcp_forced"}     = $pcp_forced;
-    $flags{"lsmodel_flag"}   = $lsmodel_flag;
     $flags{"ldas_flag"}      = $ldas_flag;
+    $flags{"lsmodel_flag"}   = $lsmodel_flag;
+    $flags{"pcp_forced"}     = $pcp_forced;
+
```
Same as above ... Are we intentionally changing the order of the two lines here (relative to that in develop)?  Unless there's a reason for the revised order, we shouldn't modify develop

saraqzhang replied:
the order of the two lines are not changed in the code, not sure why it is indicated here.

gmao-rreichle replied:
the order of the lines does seem changed, see comment above

saraqzhang replied:
same as above

gmao-rreichle commented on src/Applications/NCEP_Etc/NCEP_enkf/scripts/gmao/etc/HISTAENS.rc.tmpl:
```diff
@@ -863,3 +865,47 @@ COLLECTIONS: 'bkg.eta'
+  bkg.tavg1_2d_lfo.end_date:   >>>IOEDATE<<< ,
+  bkg.tavg1_2d_lfo.end_time:   >>>IOETIME<<< ,
+  bkg.tavg1_2d_lfo.fields:     'SLRSF'   , 'SOLAR'    , 'SWGDN'    ,
+                                'SWLAND'  , 'SURFACE'  ,
```
I'm pretty sure we do not need the **net** SW (SWLAND) in the lfo tavg file. It's a remnant from way back when.  I recommend removing the SWLAND variable and trying a single cycle without it to verify

saraqzhang replied:
if  we remove SWLAND from bkg.*lfo* in HISTAENS.rc,  should we also remove it from bkg.*lfo* and "lfo_Nx*  in  central HISTORY.rc.tmpl ( the file in gcm_app)

gmao-rreichle replied:
yes, SWLAND should be removed from both the deterministic and ensemble HISTORY files

gmao-rreichle commented on src/Applications/GEOSdas_App/fvsetup:
```diff
@@ -3499,42 +3496,54 @@ sub set_newradbc {
 #=======================================================================
 sub set_ldasANA {
+    my ($dflt, $ans1, $ans2);
+    my (@lfo_files);
```
I recognize `lfo_files` is just a variable name, but for consistency it may be better to use `bkglfo_files`

saraqzhang replied:
will modify the name.

gmao-rreichle commented on src/Applications/GEOSdas_App/fvsetup:
```diff
@@ -10031,6 +10046,8 @@ sub init_agcm_pert_rc {
   my ($chemrc, $ensFLG);
   my ($flags, $label, $string, $vars, $cmd);
+  my ($fname, $fname_old, $expid_old, $target);
+  my ($lfo_tarfile, @lfo_files, $target, $expid_old, $fname, $fname_old);
```
I recognize `lfo_tarfile` is just a variable name, but for consistency it may be better to use `bkglfo_tarfile`

saraqzhang replied:
same as above

gmao-rreichle commented on src/Applications/NCEP_Etc/NCEP_enkf/scripts/gmao/get_atmens_rst.pl:
```diff
@@ -24,11 +24,12 @@
     my ($dmlist, $ens, $hh3, $hms3, $label, $mem, $mfile, $mfile_new, $mm3);
     my ($pid, $pwd, $rdnperts_dates3_txt, $tarfile, $tarpath);
     my ($three_hr_sec, $ymd3, $yyyy3, @tarList, @tarListA);
+    my (@labels, $bkglfo_present, $elfo );
```
I recognize `elfo` is just a variable name, but for consistency it may be better to use `ebkglfo`.  
Similarly, maybe replace  `bkglfo_present` with `ebkglfo_present`

saraqzhang replied:
same as above

gmao-rreichle reviewed (commented):
@saraqzhang, I added a few inline comments.  I might be wrong on some (or even all of them), but take a look and see if any of the suggested changes make sense

saraqzhang committed:
minor edits in fvsetup and get_atmens_rst.pl
	modified:   src/Applications/GEOSdas_App/fvsetup
	modified:   src/Applications/NCEP_Etc/NCEP_enkf/scripts/gmao/get_atmens_rst.pl

gmao-rreichle committed:
Merge branch 'develop' into feature/saraqzhang/updatesetup4ladas

gmao-rreichle committed:
clarified comment re. LDAS setup (fvsetup)

gmao-rreichle committed:
fixed typo in error message (fvsetup)

saraqzhang committed:
remove SWLAND from lfo collection in HISTAENS.rc.tmpl

saraqzhang committed:
minor edit to bkg.lfo file name
	modified:   fvsetup

saraqzhang committed:
minor edits for bkg.lfo  name

saraqzhang committed:
reinstate updates for ldas flag
	modified:   AGCMrc.pm

saraqzhang committed:
edits on bkg lfo query and tar

saraqzhang committed:
reinstate member tag to bkg.lfo filename in HISTAENS.rc.tmpl

gmao-rreichle committed:
Merge branch 'develop' into feature/saraqzhang/updatesetup4ladas

saraqzhang committed:
update tags of ldasGC and gcm_App
	modified:   components.yaml

gmao-rreichle committed:
Merge branch 'develop' into feature/saraqzhang/updatesetup4ladas

gmao-rreichle commented:
@saraqzhang, @gmao-jstassi, @rtodling : I see that fvsetup maintains lists of files for the ADAS components.  E.g., for chemistry: https://github.com/GEOS-ESM/GEOSadas/blob/a4e6c0c933130bd13c294f6dd4d25f9a37108a81/src/Applications/GEOSdas_App/fvsetup#L922
Looking towards improving the LDAS integration going forward, do we need to think about having such a list for the LDAS files?  For example, LDAS needs nml input files that define the land model perturbations and the obs that it assimilates.   (These files are in plain vanilla Fortran nml format, but they're essentially rc files.)

rtodling committed:
Merge branch 'develop' into feature/saraqzhang/updatesetup4ladas

rtodling commented:
I am working on this ... please don't make changes anymore @saraqzhang @gmao-jstassi

rtodling committed:
back to the old; zero diff

rtodling committed:
identation

rtodling committed:
adjust gcmapp tag

rtodling reviewed (approved):
I changed references to older architectures in fvsetup to what they were ... I will make sure those get changed ( and removed, as this originally attempted) but in a general wya - w/ the *input files changing accordingly.
